src/Microsoft.ML/Data/TextLoader.cs

```diff
@@ -0,0 +1,116 @@
+            for (int index = 0; index < fields.Length; index++)
+            {
+                var field = fields[index];
+                var mappingAttr = field.GetCustomAttribute<ColumnAttribute>();
+                if (mappingAttr == null)
+                    throw Contracts.Except($"{field.Name} is missing ColumnAttribute");
+
+                if (Regex.Match(mappingAttr.Ordinal, @"[^(0-9,\*\-~)]+").Success)
+                    throw Contracts.Except($"{mappingAttr.Ordinal} contains invalid characters. " +
+                        $"Valid characters are 0-9, *, - and ~");
+
+                var name = mappingAttr.Name ?? field.Name;
+                if (name.Any(c => !Char.IsLetterOrDigit(c)))
+                    throw Contracts.Except($"{name} is not alphanumeric.");
+
+                if(separator != null)
+                {
+                    if(separator != "space" && separator != "tab" && separator != "comma" && separator.Length > 1)
+                        throw Contracts.Except($"{nameof(separator)} can only be one of the following: space, tab, comma" +
+                            $" or a single character.");
+                }
+
+                var col = Runtime.Data.TextLoader.Column.Parse(
+                    $"{name}:" +
+                    $"{TypeToName(field.FieldType.IsArray ? field.FieldType.GetElementType() : field.FieldType)}:" +
+                    $"{mappingAttr.Ordinal}"
+                    );
+                
+                TextLoaderColumn tlc = new TextLoaderColumn();
+                if (col.KeyRange != null)
+                {
+                    tlc.KeyRange = new KeyRange();
+                    tlc.KeyRange.Min = col.KeyRange.Min;
+                    tlc.KeyRange.Max = col.KeyRange.Max;
+                }
+
+                tlc.Name = col.Name;
+                tlc.Source = new TextLoaderRange[col.Source.Length];
+                for (int indexLocal = 0; indexLocal < tlc.Source.Length; indexLocal++)
+                {
+                    tlc.Source[indexLocal] = new TextLoaderRange
+                    {
+                        AllOther = col.Source[indexLocal].AllOther,
+                        AutoEnd = col.Source[indexLocal].AutoEnd,
+                        ForceVector = col.Source[indexLocal].ForceVector,
+                        VariableEnd = col.Source[indexLocal].VariableEnd,
+                        Max = col.Source[indexLocal].Max,
+                        Min = col.Source[indexLocal].Min
+                    };
+                }
+
+                tlc.Type = col.Type;
+                Arguments.Column[index] = tlc;
+            }
+
+            Arguments.HasHeader = useHeader;
+            Arguments.Separator = separator;
+            Arguments.AllowQuoting = allowQuotedStrings;
+            Arguments.AllowSparse = supportSparse;
+            Arguments.TrimWhitespace = trimWhitespace;
+        }
+
+        private string TypeToName(Type type)
```

Comment:
It isn't *quite* the same, since in this case `double` is mapping to `R4`, while `DvText` maps to `TX` and `DvBool` maps to `BL` (while `string` and `bool` do not map to anything, since they are not "blessed" kinds).

---
In reply to: [186311810](https://github.com/dotnet/machinelearning/pull/38#discussion_r186311810) [](ancestors = 186311810)

Reply:
Ah... I see `TlcModule.GetDataType`. That does the mappings listed here, plus a few more.

Even if for some reason we cannot use that, let's have this method use `DataKind` directly. We are specifying these things like `"R4"` and `"BL"`, but these are not random string literals, these are the names of enums in our codebase, so we should just use the enums.

---
In reply to: [186311810](https://github.com/dotnet/machinelearning/pull/38#discussion_r186311810) [](ancestors = 186311810)

Reply:
Thanks but it seems we need to use TryGetDataKind.

---
In reply to: [186565505](https://github.com/dotnet/machinelearning/pull/38#discussion_r186565505) [](ancestors = 186565505,186311810)